src/core/connection_pool.c

```diff
@@ -0,0 +1,521 @@
+
+    if (Config == NULL || ConnectionPool == NULL || Config->Registration == NULL ||
+        Config->Configuration == NULL|| Config->NumberOfConnections == 0 ||
+        Config->Handler == NULL || Config->ServerName == NULL || Config->ServerPort == 0 ||
+        (Config->Family != QUIC_ADDRESS_FAMILY_UNSPEC &&
+            Config->Family != QUIC_ADDRESS_FAMILY_INET &&
+            Config->Family != QUIC_ADDRESS_FAMILY_INET6) ||
+        (((QUIC_CONFIGURATION*)Config->Configuration)->SecurityConfig == NULL) ||
+        (Config->CibirIds != NULL && Config->CibirIdLength == 0) ||
+        (Config->CibirIds == NULL && Config->CibirIdLength != 0)) {
+        Status = QUIC_STATUS_INVALID_PARAMETER;
+        QuicTraceLogError(
+            ConnPoolInvalidParam,
+            "[conp] Invalid parameter, 0x%x",
+            Status);
+        return Status;
+    }
+
+    const size_t ServerNameLength = strnlen(Config->ServerName, QUIC_MAX_SNI_LENGTH + 1);
+    if (ServerNameLength == QUIC_MAX_SNI_LENGTH + 1) {
+        Status = QUIC_STATUS_INVALID_PARAMETER;
+        QuicTraceLogError(
+            ConnPoolServerNameTooLong,
+            "[conp] ServerName too long (%llu bytes)",
+            ServerNameLength);
+        return Status;
+    }
+
+    QuicTraceEvent(
+        ApiEnter,
+        "[ api] Enter %u (%p).",
+        QUIC_TRACE_API_CONNECTION_POOL_CREATE,
+        Config->Registration);
+
+    CxPlatZeroMemory(ConnectionPool, sizeof(HQUIC) * Config->NumberOfConnections);
+
+    //
+    // Resolve the server name or use the remote address.
+    //
+    QUIC_ADDR ResolvedRemoteAddress = { 0 };
+    QuicAddrSetFamily(&ResolvedRemoteAddress, Config->Family);
+    if (Config->ServerAddress != NULL) {
+         ResolvedRemoteAddress = *Config->ServerAddress;
+    } else {
+        Status =
+            CxPlatDataPathResolveAddress(
+                MsQuicLib.Datapath,
+                Config->ServerName,
+                &ResolvedRemoteAddress);
+        if (QUIC_FAILED(Status)) {
+            QuicTraceLogError(
+                ConnPoolResolveAddress,
+                "[conp] Failed to resolve address, 0x%x",
+                Status);
+            goto Error;
+        }
+    }
+
+    QuicAddrSetPort(&ResolvedRemoteAddress, Config->ServerPort);
+
+    //
+    // Get the local address and a port to start from.
+    //
+    QUIC_ADDR LocalAddress;
+    Status = QuicConnPoolGetStartingLocalAddress(&ResolvedRemoteAddress, &LocalAddress);
+    if (QUIC_FAILED(Status)) {
+        QuicTraceLogError(
+            ConnPoolGetLocalAddress,
+            "[conp] Failed to get local address, 0x%x",
+            Status);
+        goto Error;
+    }
+
+    uint32_t InterfaceIndex;
+    Status = QuicConnPoolGetInterfaceIndexForLocalAddress(&LocalAddress, &InterfaceIndex);
+    if (QUIC_FAILED(Status)) {
+        goto Error;
+    }
+
+    Status = CxPlatDataPathRssConfigGet(InterfaceIndex, &RssConfig);
+    if (QUIC_FAILED(Status)) {
+        QuicTraceLogError(
```

Comment:
In general, you don't need to log an error if the function itself already logs an error.